src/sync/sync_permission.cpp

```diff
@@ -0,0 +1,218 @@
+    realm->begin_transaction();
+    object_wrapper->object = Object::create<util::Any>(&context, realm, *realm->schema().find("PermissionChange"), AnyDict{
+        { "id",         util::uuid_string() },
+        { "createdAt",  Timestamp(0, 0) },
+        { "updatedAt",  Timestamp(0, 0) },
+        { "userId",     permission.condition.user_id },
+        { "realmUrl",   realm_url },
+        { "mayRead",    permission.access != Permission::AccessLevel::None },
+        { "mayWrite",   permission.access == Permission::AccessLevel::Write || permission.access == Permission::AccessLevel::Admin },
+        { "mayManage",  permission.access == Permission::AccessLevel::Admin },
+    }, false);
+    realm->commit_transaction();
+
+    auto block = [object_wrapper, callback=std::move(callback)](CollectionChangeSet, std::exception_ptr ex) mutable {
+        Object& obj = object_wrapper->object;
+        if (ex) {
+            callback(ex);
+            object_wrapper.reset();
+            return;
+        }
+        CppContext context;
+        auto status_code = obj.get_property_value<util::Any>(&context, "statusCode");
+        if (status_code.has_value()) {
+            auto code = any_cast<long long>(status_code);
+            std::exception_ptr exc_ptr = nullptr;
+            if (code) {
+                auto status = obj.get_property_value<util::Any>(&context, "statusMessage");
+                std::string error_str = (status.has_value()
+                                         ? any_cast<std::string>(status)
+                                         : std::string("Error code: ") + util::to_string(code));
+                exc_ptr = std::make_exception_ptr(PermissionChangeException(error_str, code));
+            }
+            callback(exc_ptr);
+            object_wrapper.reset();
+        }
+    };
+    object_wrapper->token = object_wrapper->object.add_notification_block(std::move(block));
+}
+
+void Permissions::delete_permission(std::shared_ptr<SyncUser> user,
+                                    Permission permission,
+                                    PermissionChangeCallback callback,
+                                    const ConfigMaker& make_config)
+{
+    permission.access = Permission::AccessLevel::None;
+    set_permission(std::move(user), std::move(permission), std::move(callback), make_config);
+}
+
+SharedRealm Permissions::management_realm(std::shared_ptr<SyncUser> user, const ConfigMaker& make_config)
+{
+    // FIXME: maybe we should cache the management Realm on the user, so we don't need to open it every time.
+    const auto realm_url = std::string("realm") + user->server_url().substr(4) + "/~/__management";
+    Realm::Config config = make_config(user, std::move(realm_url));
+    config.sync_config->stop_policy = SyncSessionStopPolicy::Immediately;
+    config.schema = Schema{
+        { "PermissionChange", {
+            { "id",             PropertyType::String,   "", "", true, true, false   },
```

Comment:
These boolean literals are impossible to understand without referring to the type's constructor (and the type isn't even named here, making that harder too).